src/index.js

```diff
@@ -1,108 +1,120 @@
+      }).attach()
+    }
+  }
+
   const styled = (
-    tagOrStyledElement: tagOrStyledElementTypeype,
-    ownStyles: Object
+    tagNameOrStyledElement: TagNameOrStyledElementType,
+    ownStyle: ComponentStyleType
   ): StyledElementType => {
-    const {tag, styles}: StyledElementAttrsType = typeof tagOrStyledElement === 'string'
-      ? {tag: tagOrStyledElement, styles: {}}
-      : tagOrStyledElement
+    const {tagName, style}: StyledElementAttrsType = typeof tagNameOrStyledElement === 'string'
+      ? {tagName: tagNameOrStyledElement, style: {}}
+      : tagNameOrStyledElement
 
-    const elementStyles = {...styles, ...ownStyles}
-    const dynamicStyles = getDynamicStyles(elementStyles)
-    const staticTag = `${tag}-${++counter}`
+    const elementStyle = {...style, ...ownStyle}
+    const dynamicStyle = getDynamicStyles(elementStyle)
+    const staticTagName = getScopedTagName(tagName)
 
     return class StyledElement extends PureComponent {
-      static tag = tag
-
-      static styles = elementStyles
+      static tagName: string = tagName
+      static style: ComponentStyleType = elementStyle
 
       props: StyledElementPropsType
 
-      tagScoped = ''
+      dynamicTagName = ''
 
       constructor(props) {
         super(props)
-        this.tagScoped = `${tag}-${++counter}`
+        this.dynamicTagName = getScopedTagName(tagName)
       }
 
       componentWillMount() {
-        if (!sheet) {
-          sheet = jss.createStyleSheet(baseStyles, {
-            link: true,
-            meta: 'StaticBaseSheet',
-          }).attach()
-
-          dynamicSheet = jss.createStyleSheet({}, {
-            link: true,
-            meta: 'DynamicComponentSheet',
-          }).attach()
-        }
+        mountSheets()
 
-        if (!sheet.getRule(staticTag)) {
-          sheet.addRule(staticTag, elementStyles)
+        if (!sheets.staticSheet.getRule(staticTagName)) {
+          sheets.staticSheet.addRule(staticTagName, elementStyle)
         }
 
-        if (dynamicStyles && !dynamicSheet.getRule(this.tagScoped)) {
-          dynamicSheet
+        if (dynamicStyle && !sheets.dynamicSheet.getRule(this.dynamicTagName)) {
+          sheets.dynamicSheet
             .detach()
-            .addRule(this.tagScoped, dynamicStyles)
-          dynamicSheet
-            .update(this.tagScoped, this.props)
+            .addRule(this.dynamicTagName, dynamicStyle)
+          sheets.dynamicSheet
+            .update(this.dynamicTagName, this.props)
             .attach()
             .link()
         }
       }
 
       componentWillReceiveProps(nextProps: StyledElementPropsType) {
-        if (dynamicStyles) {
-          dynamicSheet.update(this.tagScoped, nextProps)
+        if (dynamicStyle) {
+          sheets.dynamicSheet.update(this.dynamicTagName, nextProps)
         }
       }
 
       componentWillUnmount() {
-        dynamicSheet.deleteRule(this.tagScoped)
+        sheets.dynamicSheet.deleteRule(this.dynamicTagName)
```

Comment:
why do we remove the rule, we could reuse that rule next time the component mounts, no?